zepa.go

```
@@ -15,38 +17,170 @@ const (
+	SUB
+	CMP
+	JUMP
+	LOAD
+	STORE
+	FETCH
+)
+
+const (
+	opcodeLength = 6
+	rdLength     = 5
+	rs1Length    = 5
+	rs2Length    = 5
+	funct5Length = 5
+	funct6Length = 6
+	immediateLen = 16
+	word         = 32
+)
+
+const (
+	opcodeBitMask    = 0b111111
+	registerBitMask  = 0b11111
+	immediateBitMask = 0b1111111111111111
+	funct5BitMask    = 0b11111
+	funct6BitMask    = 0b111111
 )
 
+var operations = map[byte]Operation{
+	byte(MV):    (*Machine).mv,
+	byte(ADD):   (*Machine).add,
+	byte(SUB):   (*Machine).sub,
+	byte(CMP):   (*Machine).cmp,
+	byte(JUMP):  (*Machine).jump,
+	byte(LOAD):  (*Machine).load,
+	byte(STORE): (*Machine).store,
+}
+
+type Instruction struct {
+	opcode    func(m *Machine, inst Instruction)
+	rd        Register
+	rs1       Register
+	rs2       Register
+	funct5    byte
+	funct6    byte
+	immediate uint16
+}
+
 type Machine struct {
 	memory    []byte
-	registers map[Register]byte
+	registers map[Register]uint32
 }
 
+func (m *Machine) mv(inst Instruction) {}
+
+func (m *Machine) add(inst Instruction) {}
+
+func (m *Machine) sub(inst Instruction) {}
+
+func (m *Machine) cmp(inst Instruction) {}
+
+func (m *Machine) jump(inst Instruction) {}
+
+func (m *Machine) load(inst Instruction) {}
+
+func (m *Machine) store(inst Instruction) {}
+
 func (m *Machine) fetch() {
-	currentInstructionAddress := m.registers[pc]
-	currentInstruction := m.memory[currentInstructionAddress]
-	// Update Instruction Register with current instruction
-	m.registers[ir] = currentInstruction
-	// Increment Program Counter
-	m.registers[pc] += 1
+	var completeInstruction uint32 = 0
+	for i := 0; i < 4; i++ {
+		currentInstructionAddress := m.registers[pc]
+		currentInstruction := m.memory[currentInstructionAddress]
+		completeInstruction = completeInstruction | uint32(currentInstruction)<<(24-8*i)
+		m.registers[pc] += 1
+	}
+	m.registers[ir] = completeInstruction
 }
 
-func (m *Machine) decode() {}
+func (m *Machine) decodeRTypeInst(instruction uint32) Instruction {
```

Should return an error checking the operands